### ISSUE TITLE
VMNull in add_method in indirect use

### ISSUE THREAD

**Issue by gfldex** [rakudo/rakudo#2897]
```
https://gist.github.com/gfldex/dc47e3c4981cd63543d3e20f3e6b7b5e

#  warn-in-add_method.pm6 
sub EXPORT {
    Nil.^add_method('list', method ($self: ){ my $i = 41++; Empty });
    Nil.^compose;
    %()
}
# EOF

# warn-in-add_method-indirect.pm6
use warn-in-add_method;
# EOF

#  warn-in-add_method.p6 
use v6.d;

use lib '.';
use warn-in-add_method-indirect;

Nil.list();
# EOF
```
Yields:
```
Cannot invoke this object (REPR: Null; VMNull)
  in method <anon> at /home/dex/projects/perl6/rakudobug/warn-in-add_method.pm6 (warn-in-add_method) line 2
  in block <unit> at warn-in-add_method.p6 line 6
```

It doesn't really matter what the first statement in the added method is as long as it's not optimised away. The indirection is needed to trigger this and the `sub EXPORT´ is required to avoid STable-confusion (what may also be a bug).

$ perl6 --version
This is Rakudo version 2019.03.1-292-g48a12551f built on MoarVM version 2019.03-110-g081823ffc
implementing Perl 6.d.

**Comment by vrurg** [comment]
Golfed the code down to:

_Bar.pm6_:
```perl6
sub sub-producer is export {
    my $i = 0;
    sub { ++$i }
}
```

_Foo.pm6_:
```perl6
use Bar;
our &foo;
BEGIN &foo = sub-producer;
say "foo: ", foo;
```

foo.p6
```perl6
use lib '.';
use Foo;
```

**Comment by vrurg** [comment]
Reopening until the next NQP bump.

**Comment by vrurg** [comment]
Closed via MoarVM/MoarVM#1209 and 7e3d023b4b070ed798915336a98f2bbc7fbac2e7

**Comment by vrurg** [comment]
Closing via MoarVM/MoarVM#1209 and 7e3d023b4b070ed798915336a98f2bbc7fbac2e7
